processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java

```diff
@@ -176,26 +246,58 @@ public int getNumRows()
 
     private List<ByteBuffer> updateValues(
         ByteBuffer key,
-        List<DimensionSelector> dims
+        List<DimensionSelector> dims,
+        final List<String> dimNames,
+        final List<DimensionHandler> dimHandlers
     )
     {
       if (dims.size() > 0) {
         List<ByteBuffer> retVal = null;
         List<ByteBuffer> unaggregatedBuffers = null;
 
         final DimensionSelector dimSelector = dims.get(0);
-        final IndexedInts row = dimSelector.getRow();
-        if (row == null || row.size() == 0) {
-          ByteBuffer newKey = key.duplicate();
-          newKey.putInt(dimSelector.getValueCardinality());
-          unaggregatedBuffers = updateValues(newKey, dims.subList(1, dims.size()));
+        final DimensionHandler dimHandler = dimHandlers.get(0);
+        final ColumnCapabilities capabilities = dimSelector.getDimCapabilities();
+        IndexedInts intVals;
+        ValueType type;
+
+        if (dimHandler != null) {
+          type = capabilities.getType();
         } else {
-          for (Integer dimValue : row) {
-            ByteBuffer newKey = key.duplicate();
-            newKey.putInt(dimValue);
-            unaggregatedBuffers = updateValues(newKey, dims.subList(1, dims.size()));
-          }
+          // Dimension does not exist, dimSelector is a NullDimensionSelector.
+          // The null value WILL appear in the result rows.
+          // tests in GroupByQueryRunnerTest expect this behavior. (e.g., testGroupByWithMetricColumnDisappears())
+          type = ValueType.STRING;
         }
+
+        switch (type) {
```

Comment:
Doing it as a switch/case is going to eliminate the ability for new column types to be added as an extension.  Maybe not the end of the world, but if it's possible to get it behind the interfaces, then extensions become a reality.